interface/src/hooks/useScrollRestoration.tsx

```diff
@@ -0,0 +1,99 @@
+import React from 'react';
+import { useLocation } from 'react-router-dom';
+
+interface ScrollPositionProviderProps {
+  children: React.ReactNode;
+}
+
+interface IScrollPositionContext {
+  saveScrollPosition: (key: string, position: number) => void;
+  getScrollPosition: (key: string) => number;
+}
+
+const ScrollPositionContext = React.createContext<
+  IScrollPositionContext | undefined
+>(undefined);
+
+function useScrollPosition(): IScrollPositionContext {
+  const context = React.useContext(ScrollPositionContext);
+  if (!context) {
+    throw new Error(
+      'useScrollPosition must be used within a ScrollPositionProvider',
+    );
+  }
+  return context;
+}
+
+export function ScrollPositionProvider({
```

Comment:
This entire thing is fantastic, I need to play around with it for a bit. But I really love this implementation.